Index: libsl1550/libsecondlife/_Packets_.cs
===================================================================
--- libsl1550/libsecondlife/_Packets_.cs	(revision 33)
+++ libsl1550/libsecondlife/_Packets_.cs	(working copy)
@@ -782,7 +782,7 @@
         public abstract void FromBytes(byte[] bytes, ref int i, ref int packetEnd, byte[] zeroBuffer);
         public abstract void FromBytes(Header header, byte[] bytes, ref int i, ref int packetEnd, byte[] zeroBuffer);
         public int ResendCount;
-        internal int TickCount;
+        public int TickCount;
 
         public abstract byte[] ToBytes();
         public static PacketType GetType(ushort id, PacketFrequency frequency)
