From 8957d854bee84942033647a09e13d7f3681f17aa Mon Sep 17 00:00:00 2001
From: Robert Adams <Robert.Adams@intel.com>
Date: Mon, 26 Mar 2012 11:05:02 -0700
Subject: [PATCH] Call setWorldTransform when object is going inactive

---
 .../Dynamics/btDiscreteDynamicsWorld.cpp           |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/src/BulletDynamics/Dynamics/btDiscreteDynamicsWorld.cpp b/src/BulletDynamics/Dynamics/btDiscreteDynamicsWorld.cpp
index 954ef24..2767899 100755
--- a/src/BulletDynamics/Dynamics/btDiscreteDynamicsWorld.cpp
+++ b/src/BulletDynamics/Dynamics/btDiscreteDynamicsWorld.cpp
@@ -640,6 +640,9 @@ void	btDiscreteDynamicsWorld::updateActivationState(btScalar timeStep)
 					{
 						body->setAngularVelocity(btVector3(0,0,0));
 						body->setLinearVelocity(btVector3(0,0,0));
+						// when sleeping, force the motion state to be called/updated
+						if (body->getMotionState())
+							body->getMotionState()->setWorldTransform(body->getWorldTransform());
 					}
 
 				}
-- 
1.7.9

