From 2a2bb8512bc0b7c804b30f5bf8eaf10ea721eb14 Mon Sep 17 00:00:00 2001
From: Robert Adams <Robert.Adams@intel.com>
Date: Sun, 27 Jan 2013 21:07:13 -0800
Subject: [PATCH 2/2] Disable internal profile gathering

---
 src/LinearMath/btQuickprof.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/LinearMath/btQuickprof.h b/src/LinearMath/btQuickprof.h
index 8691400..a688912 100755
--- a/src/LinearMath/btQuickprof.h
+++ b/src/LinearMath/btQuickprof.h
@@ -16,7 +16,7 @@
 #define BT_QUICK_PROF_H
 
 //To disable built-in profiling, please comment out next line
-//#define BT_NO_PROFILE 1
+#define BT_NO_PROFILE 1
 #ifndef BT_NO_PROFILE
 #include <stdio.h>//@todo remove this, backwards compatibility
 #include "btScalar.h"
-- 
1.7.9

