diff -r 6057b1cda03d libsecondlife/AssetManager.cs
--- a/libsecondlife/AssetManager.cs	Wed Mar 12 22:33:53 2008 +0000
+++ b/libsecondlife/AssetManager.cs	Mon Apr 07 16:25:09 2008 +0200
@@ -31,7 +31,7 @@
         Clothing = 5,
         /// <summary>Primitive that can contain textures, sounds, 
         /// scripts and more</summary>
-        Primitive = 6,
+        Object = 6,
         /// <summary>Notecard asset</summary>
         Notecard = 7,
         /// <summary>Holds a collection of inventory items</summary>
@@ -668,7 +668,7 @@
                 case AssetType.Texture:
                     asset = new AssetTexture();
                     break;
-                case AssetType.Primitive:
+                case AssetType.Object:
                     asset = new AssetPrim();
                     break;
                 case AssetType.Clothing:
diff -r 6057b1cda03d libsecondlife/AssetTypes.cs
--- a/libsecondlife/AssetTypes.cs	Wed Mar 12 22:33:53 2008 +0000
+++ b/libsecondlife/AssetTypes.cs	Mon Apr 07 16:25:09 2008 +0200
@@ -167,7 +167,7 @@
 
     public class AssetPrim : Asset
     {
-        public override AssetType AssetType { get { return AssetType.Primitive; } }
+        public override AssetType AssetType { get { return AssetType.Object; } }
 
         public AssetPrim() { }
 
