Index: libsecondlife/libsecondlife.build
===================================================================
--- libsecondlife/libsecondlife.build	(revision 31)
+++ libsecondlife/libsecondlife.build	(working copy)
@@ -67,7 +67,7 @@
 		</exec>
 	</target>
 
-	<target name="build" depends="init map" description="compiles the source code">
+	<target name="build" depends="init" description="compiles the source code">
 	    <if test="${property::exists('CCNetLabel')}">
 			<!-- Produce AssemblyInfo.cs for libsecondlife.dll -->
 			<asminfo output="AssemblyInfo.cs" language="CSharp">
